src/nvim/os/win_defs.h

```diff
@@ -1,25 +1,88 @@
 #ifndef NVIM_OS_WIN_DEFS_H
 #define NVIM_OS_WIN_DEFS_H
 
+#include "auto/config.h"
+
+#ifndef LC_MESSAGES
+// FIXME: this is a hack - also used in ex_cmds2.c
+// to build in windows without libintl
+# define LC_MESSAGES LC_COLLATE
+#endif
+
+// winsock2.h must be before windows.h - or so says Mingw
+#include <winsock2.h>
 #include <windows.h>
+// We have our own RGB macro in macros.h
+#undef RGB
+
+#include <io.h>
+#include <uv.h>
+
+#include <stdio.h>
+#include <time.h>
+#include <sys/stat.h>
+
+// FIXME: this is only used by if_cscope.c
+#ifndef S_ISLNK
+# define S_ISLNK(mode) 0
+#endif
+
+// For MSVC
+#ifdef _MSC_VER
+# ifndef restrict
+#  define restrict __restrict
+# endif
+# ifndef inline
+#  define inline __inline
+# endif
+# ifndef __func__
+#  define __func__ __FUNCTION__
+# endif
+# ifndef STDOUT_FILENO
+#  define STDOUT_FILENO _fileno(stdout)
+# endif
+# ifndef STDERR_FILENO
+#  define STDERR_FILENO _fileno(stderr)
+# endif
+# ifndef W_OK
+// There is no W_OK for MSVC but it is 02
+# define W_OK 2
+# endif
+# ifndef S_ISDIR
+#  define S_ISDIR(mode) (((mode) & S_IFMT) == S_IFDIR)
+# endif
+# ifndef S_ISREG
+#  define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
+# endif
+# ifndef S_IXUSR
+// #  define S_IXUSR(mode) (((mode) & S_IFMT) == S_IEXEC)
+#   define S_IXUSR S_IEXEC
+# endif
+
+// FIXME: (equalsraf)
+// This can be removed for MSVC2015
+# if defined(_MSC_VER) && _MSC_VER < 1900
+#  define snprintf _snprintf
+# endif
+
+# ifndef SSIZE_MAX
+#  if defined(_WIN64)
+#   define SSIZE_MAX _I64_MAX
+#  else
+#   define SSIZE_MAX LONG_MAX
+#  endif
+# endif
```

Comment:
Rebasing won't remove these items, these should be explicitly removed from this PR.